controllers/configmap_controller.go

```diff
@@ -18,45 +18,229 @@ package controllers
+			return nil
+		}
+		// if the address parses into an IP but is not ipv4 it must be ipv6
+		return fmt.Errorf("ipv6 is not supported")
+	}
+	// Do a check that the DNS provided is valid
+	addressList, err := net.LookupHost(address)
+	if err != nil {
+		return errors.Wrapf(err, "error looking up DNS")
+	}
+	if len(addressList) == 0 {
+		return fmt.Errorf("DNS did not resolve to an address")
+	}
+	return nil
+}
+
+// addAndRemoveNodes corrects the discrepancy between the "expected" hosts slice, and the "actual" nodelist
+func (r *ConfigMapReconciler) addAndRemoveNodes(ctx context.Context, instances *core.ConfigMap) error {
+	// Get the list of instances that are expected to be Nodes
+	hosts, err := r.parseHosts(instances.Data)
+	if err != nil {
+		return errors.Wrapf(err, "unable to parse hosts from configmap")
+	}
+
+	nodes := &core.NodeList{}
+	if err := r.client.List(ctx, nodes); err != nil {
+		return errors.Wrap(err, "error listing nodes")
+	}
+
+	// For each host, ensure that it is configured into a node. On error of any host joining, return error and requeue.
+	// It is better to return early like this, instead of trying to configure as many nodes as possible in a single
+	// reconcile call, as it simplifies error collection. The order the map is read from is psuedo-random, so the
+	// configuration effort for configurable hosts will not be blocked by a specific host that has issues with
+	// configuration.
+	for _, host := range hosts {
+		err := r.ensureNodeExists(host, nodes)
+		if err != nil {
+			r.recorder.Eventf(instances, core.EventTypeWarning, "InstanceSetupFailure",
+				"unable to join instance with address %s to the cluster", host.address)
+			return errors.Wrapf(err, "error configuring host with address %s", host.address)
+		}
+	}
+	// TODO: Remove BYOH nodes that are not associated with the parsed host list
+	//       https://issues.redhat.com/browse/WINC-582
+
+	// Once all the proper Nodes are in the cluster, configure the prometheus endpoints.
+	if err := r.prometheusNodeConfig.Configure(); err != nil {
+		return errors.Wrap(err, "unable to configure Prometheus")
+	}
+	return nil
+}
+
+// ensureNodeExists ensures that the given host has an associated Node
+func (r *ConfigMapReconciler) ensureNodeExists(instance *instanceInfo, nodes *core.NodeList) error {
+	_, found := findNode(instance.address, nodes)
+	if found {
+		// TODO: Remove and re-add the node if it is not 'Ready'
+		// TODO: Check version for upgrade case https://issues.redhat.com/browse/WINC-580
+		return nil
+	}
+
+	if err := r.configureInstance(instance, "", map[string]string{BYOHAnnotation: "true"}); err != nil {
+		return errors.Wrap(err, "error configuring node")
+	}
+
+	return nil
+}
 
-	return ctrl.Result{}, nil
+// findNode returns a pointer to the node with an address matching the given address and a bool indicating if the node
+// was found or not.
+func findNode(address string, nodes *core.NodeList) (*core.Node, bool) {
```

Comment:
looks like the node returned here is not used anywhere, in that case the function can return only a bool